ddtrace/internal/writer/writer.py

```diff
@@ -376,47 +376,52 @@ def flush_queue(self, raise_exc: bool = False):
             self._set_drop_rate()
 
     def _flush_queue_with_client(self, client: WriterClientBase, raise_exc: bool = False) -> None:
-        n_traces = len(client.encoder)
-        try:
-            encoded, n_traces = client.encoder.encode()
+        total_traces = len(client.encoder)
+        encoded_traces = total_traces  # Declare the variable in case encode() fails
+        sent_traces = 0
+        while total_traces > sent_traces:
```

Comment:
I don't think this is a super obvious control flow.

That is, it isn't clear that `client.encoder.encode()` may return only a subset of encoded data. and `len(client.encoder)` being different from the length given from `client.encoder.encode()` is also a bit weird (IMHO)

Would it be better to change `encoder.encode() -> List[(bytes, n_traces)]`, so the existing encoders will just return 1 payload and the number of traces, and the ci visibility one can return multiple payloads, then we loop through each payload and try to send?


I worry with the way this is designed, if we are unable to communicate with the agent (e.g. no agent running) then this is probably going to be an infinite loop?

We only append to `sent_traces` if the sending to the agent ultimately succeeds.